source/margarita/states.rst

```diff
@@ -0,0 +1,621 @@
+of which is a public SSH key for that user.  (Paste it in from their
+public key file, one long line.)  E.g.::
+
+    users:
+      user1:
+        public_key:
+          - ssh-rsa ADFSDFSDFDFSDFSDF....DFSDFSDFSDF
+      user2:
+        public_key:
+          - ssh-rsa DFSUDFJSDJFSDJKF...SDFJSDKFSDF
+
+Dependencies:
+
+- :ref:`users.groups`
+
+.. _project.dirs:
+
+project.dirs
+~~~~~~~~~~~~
+
+Arrange for the project's main directories to be created, e.g.
+``/var/www/<project_name>``,
+``/var/www/<project_name>/log``,
+``/var/www/<project_name>/ssh``, and
+``/var/www/<project_name>/services``.
+
+Directories are owned by the project user.
+
+Dependencies:
+
+- :ref:`project.user`
+
+.. _project.django:
+
+project.django
+~~~~~~~~~~~~~~
+
+Creates a ``manage.sh`` file in the project directory that invokes
+the Django management tool with the right settings.
+
+Depends on other sls files that also do Django-related setup.
+
+Dependencies:
+
+- :ref:`project.user`
+- :ref:`project.dirs`
+- :ref:`project.venv`
+
+.. _project.queue:
+
+project.queue
+~~~~~~~~~~~~~
+
+Arrange for rabbitmq server to be installed and run.
+
+Create rabbitmq user named ``<project_name>_<environment>``, with
+password BROKER_PASSWORD from secrets.
+
+Open the firewall for rabbitmq access to other :ref:`app_minions <minions>` servers.
+
+Pillar configuration:
+
+* ``secrets:BROKER_PASSWORD``: The password to set on the rabbitmq user.
+
+Dependencies:
+
+- :ref:`rabbitmq`
+- :ref:`ufw`
+
+.. _project.repo:
+
+project.repo
+~~~~~~~~~~~~
+
+Checks out the appropriate version of the project source code to
+``/var/www/<project_name>/source``.  Or if environment is ``local``,
+rsyncs from the current local directory to the source dir on vagrant.
+
+Create the ``/var/www/<project_name>/source/.env`` file containing all
+the environment settings needed to run the project.
+
+Create a wrapper script ``/var/www/<project_name>/source/dotenv.sh``
+that sets up the environment from ``.env`` then runs another command.
+E.g.::
+
+    cd /var/www/project
+    source/dotenv.sh env/bin/python source/manage.py shell
+
+Pillar configuration:
+
+* ``github_deploy_key`` (string): Optional, contains text of the Github deploy key
+  to use to access the repository.
+* ``repo:url`` (string): Git repository URL
+* ``repo:branch`` (string): Branch to check out. Optional; default is ``master``.
+
+Dependencies:
+
+- :ref:`project.dirs`
+- :ref:`project.user`
+- :ref:`version-control`
+- :ref:`sshd.github`
+
+.. _project.user:
+
+project.user
+~~~~~~~~~~~~
+
+Create a local user named ``<project_name>`` and add it to the
+``www-data`` group.
+
+.. _project.venv:
+
+project.venv
+~~~~~~~~~~~~
+
+Create a virtualenv for the project (at ``/var/www/<project_name>/env``)
+and install Python requirements listed in
+``/var/www/<project_name>/source/requirements/dev.txt`` if the
+environment is ``local``, and otherwise from ``production.txt``.
+
+Also installs ``ghostscript`` ?!?!?!
```

Comment:
Looks like Pillow uses ghostscript for EPS support? https://pillow.readthedocs.org/en/stable/handbook/image-file-formats.html#eps Not sure why we need it in both locations though.